token/program-2022/src/serialization.rs

```diff
@@ -76,3 +94,174 @@ pub mod coption_fromstr {
+pub mod elgamalpubkey_fromstr {
+    use {
+        serde::{
+            de::{Error, Visitor},
+            Deserializer, Serializer,
+        },
+        solana_zk_token_sdk::zk_token_elgamal::pod::ElGamalPubkey,
+        std::fmt,
+    };
+
+    const ELGAMAL_PUBKEY_LEN: usize = 32;
+
+    /// serialize ElGamalPubkey values supporting Display trait
+    pub fn serialize<S>(x: &ElGamalPubkey, s: S) -> Result<S::Ok, S::Error>
+        where
+            S: Serializer,
+    {
+        s.serialize_str(&x.to_string())
+    }
+
+    struct ElGamalPubkeyVisitor;
+
+    impl<'de> Visitor<'de> for ElGamalPubkeyVisitor {
+        type Value = ElGamalPubkey;
+
+        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+            formatter.write_str("a FromStr type")
+        }
+
+        fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+            where
+                E: Error,
+        {
+            let array = super::base64_to_bytes::<ELGAMAL_PUBKEY_LEN, E>(v)?;
+            Ok(ElGamalPubkey(array))
+        }
+    }
+
+    /// deserialize ElGamalPubkey values from str
+    pub fn deserialize<'de, D>(d: D) -> Result<ElGamalPubkey, D::Error>
+        where
+            D: Deserializer<'de>,
+    {
+        d.deserialize_str(ElGamalPubkeyVisitor)
+    }
+}
+
+/// deserialization Visitors for local types
+pub mod visitors {
+    use {
+        crate::{
+            pod::OptionalNonZeroElGamalPubkey,
+            pod::OptionalNonZeroPubkey,
+        },
+        serde::{
+            de::{Error, Unexpected, Visitor},
+        },
+        solana_program::{pubkey::Pubkey},
+        solana_zk_token_sdk::zk_token_elgamal::pod::ElGamalPubkey,
+        std::{convert::TryFrom, fmt, str::FromStr}
+    };
+
+    /// Visitor for deserializing OptionalNonZeroPubkey
+    pub(crate) struct OptionalNonZeroPubkeyVisitor;
+
+    impl<'de> Visitor<'de> for OptionalNonZeroPubkeyVisitor {
+        type Value = OptionalNonZeroPubkey;
+
+        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+            formatter.write_str("a Pubkey in base58 or `null`")
+        }
+
+        fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+            where
+                E: Error,
+        {
+            let pkey = Pubkey::from_str(&v)
+                    .map_err(|_| Error::invalid_value(Unexpected::Str(v), &"value string"))?;
+
+            OptionalNonZeroPubkey::try_from(Some(pkey))
+                .map_err(|_| Error::custom("Failed to convert from pubkey"))
+        }
+
+        fn visit_unit<E>(self) -> Result<Self::Value, E>
+            where
+                E: Error,
+        {
+            OptionalNonZeroPubkey::try_from(None)
+                .map_err(|e| Error::custom(e.to_string()))
+        }
+    }
+
+    /// Visitor for deserializing OptionalNonZeroElGamalPubkey
+    pub(crate) struct OptionalNonZeroElGamalPubkeyVisitor;
+    const OPTIONAL_NZ_ELG_LEN: usize = 32;
```

Comment:
nit: please spell this out in its entirety, otherwise I'll always wonder what New Zealand has to do with this code
```suggestion
    const OPTIONAL_NONZERO_ELGAMAL_PUBKEY_LEN: usize = 32;
```

Or you can just call it `ELGAMAL_PUBKEY_LEN` again

Reply:
Done